strings/base_collections_base.h

```diff
@@ -189,64 +301,85 @@ WINRT_EXPORT namespace winrt
         {
-            if (index > static_cast<D const&>(*this).get_container().size())
+            static_cast<D&>(*this).perform_exclusive([&]
             {
-                throw hresult_out_of_bounds();
-            }
+                if (index > static_cast<D const&>(*this).get_container().size())
+                {
+                    throw hresult_out_of_bounds();
+                }
 
-            this->increment_version();
-            static_cast<D&>(*this).get_container().insert(static_cast<D const&>(*this).get_container().begin() + index, static_cast<D const&>(*this).wrap_value(value));
+                this->increment_version();
+                static_cast<D&>(*this).get_container().insert(static_cast<D const&>(*this).get_container().begin() + index, static_cast<D const&>(*this).wrap_value(value));
+            });
         }
 
         void RemoveAt(uint32_t const index)
         {
-            if (index >= static_cast<D const&>(*this).get_container().size())
+            static_cast<D&>(*this).perform_exclusive([&]
             {
-                throw hresult_out_of_bounds();
-            }
+                if (index >= static_cast<D const&>(*this).get_container().size())
+                {
+                    throw hresult_out_of_bounds();
+                }
 
-            this->increment_version();
-            static_cast<D&>(*this).get_container().erase(static_cast<D const&>(*this).get_container().begin() + index);
+                this->increment_version();
+                static_cast<D&>(*this).get_container().erase(static_cast<D const&>(*this).get_container().begin() + index);
+            });
         }
 
         void Append(T const& value)
         {
-            this->increment_version();
-            static_cast<D&>(*this).get_container().push_back(static_cast<D const&>(*this).wrap_value(value));
+            static_cast<D&>(*this).perform_exclusive([&]
+            {
+                this->increment_version();
+                static_cast<D&>(*this).get_container().push_back(static_cast<D const&>(*this).wrap_value(value));
+            });
         }
 
         void RemoveAtEnd()
         {
-            if (static_cast<D const&>(*this).get_container().empty())
+            static_cast<D&>(*this).perform_exclusive([&]
             {
-                throw hresult_out_of_bounds();
-            }
+                if (static_cast<D const&>(*this).get_container().empty())
+                {
+                    throw hresult_out_of_bounds();
+                }
 
-            this->increment_version();
-            static_cast<D&>(*this).get_container().pop_back();
+                this->increment_version();
+                static_cast<D&>(*this).get_container().pop_back();
+            });
         }
 
         void Clear() noexcept
         {
-            this->increment_version();
-            static_cast<D&>(*this).get_container().clear();
+            static_cast<D&>(*this).perform_exclusive([&]
```

Comment:
I get the pattern, but why call `(*this).perform_exclusive` here vs. types that return locks like `auto e = lock_exclusive()` / `auto s = lock_shared()` which return "the right type of unlocker"? Might reduce the code churn here and some compile time cost?

Reply:
There's a couple of things to note here, but ultimately it's just taste. The most obvious limitation to doing it that way requires the return type to _at least_ be move constructible. E.g `std::lock_guard` wouldn't be an option since it is non-copyable and non-movable (of course `std::unique_lock` would be a viable alternative). In this case, it would seem as though `slim_lock_guard`/`slim_shared_lock_guard` *are* copy/move constructible, but it's actually unsafe to do so; this is a bug.

You also have the issue where the function would have to return _something_ as you can't do `auto guard = fn();` for a `void` returning function, e.g. in the case of the single threaded collections. A natural alternative might be to return a null `void*`, but then you'd need to consistently decorate the assignment with `[[maybe_unused]]`.

Reply:
Because non-movable types cannot be used as the return type of a function. [Example](https://godbolt.org/z/zv44E4). `std::lock_guard` is an example from the STL. `slim_lock_guard` is designed similarly. Of course, there are alternatives. `std::unique_lock` supports "invalidation" and therefore can be moved. Similarly we can add `unique_slim_lock_guard`, etc. Now you're trading off reliance that the compiler will inline lambdas/forwarding calls - something compilers are really good at - with reliance that the compiler will optimize validation checks - something compilers are also pretty good at.

Reply:
Locks might not be movable, but lock guards can be?  Yes you'd need to use a pointer back to the original lock, or some "do the unlock" flag if you keep the reference mode.  I don't get what "optimize validation checks" has to do with this.  Ultimately the "return a guard" or "initialize a guard" models just seem more readable with no loss of functionality.

Reply:
I know it's frowned upon, but I've always preferred guards that can be returned from a function. Having said that, we'd like to remove all of the public locking primitives from C++/WinRT in favor of std alternatives.

Reply:
> I don't get what "optimize validation checks" has to do with this
Because the destructor would have to look like this:
```c++
~unique_lock()
{
    if (still_valid)
    {
        m_lock->unlock();
    }
}
```
That's the whole reason for both `lock_guard` and `unique_lock` exist. Don't pay for what you don't use

Reply:
You can return a non-movable type from a function, provided you use RVO. [Example](https://gcc.godbolt.org/z/xrjqP4). The class `NonMovable` is not movable (as evidenced by our inability to move it on line 17), but the function `g` happily returns it on line 11. Our usage pattern here would be RVO-friendly.

Reply:
Fair enough. I'm too used to assuming backwards compatibility with previous standards.